Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm

```diff
@@ -78,31 +156,50 @@
         script = mergeDictionariesAndSetValues(script, @{ key: func.toString });
     }
 
-    // FIXME: <https://webkit.org/b/259954> May need to remove unsupported keys.
-    // FIXME: Handle script injection in the UI Process.
+    WebExtensionScriptInjectionParameters parameters;
+    parseTargetInjectionOptions(script[targetKey], parameters, outExceptionString);
+    parseScriptInjectionOptions(script, parameters);
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::ScriptingExecuteScript(WTFMove(parameters)), [protectedThis = Ref { *this }, callback = WTFMove(callback)](std::optional<Vector<WebExtensionScriptInjectionResultParameters>> results) {
+        if (!results) {
+            callback->call();
+            return;
+        }
+
+        callback->call(toAPI(results.value()));
+    }, extensionContext().identifier().toUInt64());
 }
 
-void WebExtensionAPIScripting::insertCSS(NSDictionary *cssInfo, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+void WebExtensionAPIScripting::insertCSS(WebPage* page, NSDictionary *cssInfo, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
 {
     // Documentation: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/scripting/insertCSS
 
     if (!validateCSS(cssInfo, outExceptionString))
         return;
 
-    // FIXME: <https://webkit.org/b/259954> May need to remove unsupported keys.
-    // FIXME: Handle css injection in the UI Process.
+    WebExtensionScriptInjectionParameters parameters;
+    parseTargetInjectionOptions(cssInfo[targetKey], parameters, outExceptionString);
+    parseCSSInjectionOptions(cssInfo, parameters);
 
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::ScriptingInsertCSS(page->webPageProxyIdentifier(), WTFMove(parameters)), [protectedThis = Ref { *this }, callback = WTFMove(callback)]() {
+        callback->call();
+    }, extensionContext().identifier().toUInt64());
 }
 
-void WebExtensionAPIScripting::removeCSS(NSDictionary *cssInfo, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+void WebExtensionAPIScripting::removeCSS(WebPage* page, NSDictionary *cssInfo, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
```

Comment:
Ditto lack of `page` null check, or perhaps consider just passing the WebPageProxy identifier, which seem to be the only thing you need from it.

Reply:
I'll fix the code generator to pass references here instead. We do need the page sometimes, so it can't just be identifier.